python/tests/map_to_ancestors.py

```diff
@@ -0,0 +1,399 @@
+                    edges = []
+                edges.append(e)
+            self.process_parent_edges(edges)
+        self.finalise_references()
+        ts = self.tables.tree_sequence()
+        return ts, self.node_id_map
+
+    def process_parent_edges(self, edges):
+        """
+        Process all of the edges for a given parent.
+        """
+        assert len({e.parent for e in edges}) == 1
+        parent = edges[0].parent
+        S = []
+        for edge in edges:
+            x = self.A_head[edge.child]
+            while x is not None:
+                if x.right > edge.left and edge.right > x.left:
+                    y = Segment(max(x.left, edge.left), min(x.right, edge.right), x.node)
+                    S.append(y)
+                x = x.next
+        self.merge_labeled_ancestors(S, parent)
+        self.check_state()
+
+    def merge_labeled_ancestors(self, S, input_id):
+        """
+        All ancestry segments in S come together into a new parent.
+        The new parent must be assigned and any overlapping segments coalesced.
+        """
+        output_id = self.node_id_map[input_id]
+        is_sample = output_id != -1
+        if is_sample:
+            # Free up the existing ancestry mapping.
+            x = self.A_tail[input_id]
+            assert x.left == 0 and x.right == self.sequence_length
+            self.A_tail[input_id] = None
+            self.A_head[input_id] = None
+
+        is_ancestor = input_id in self.ancestors
+        prev_right = 0
+        for left, right, X in overlapping_segments(S):
+            if is_ancestor or is_sample:
+                if output_id == -1:
+                    output_id = self.record_node(input_id)
+                for x in X:
+                    ancestry_node = x.node
+                    self.record_edge(left, right, output_id, ancestry_node)
+                self.add_ancestry(input_id, left, right, output_id)
+
+                if is_sample and left != prev_right:
+                    # Fill in any gaps in the ancestry for the sample.
+                    self.add_ancestry(input_id, prev_right, left, output_id)
+
+            else:
+                for x in X:
+                    ancestry_node = x.node
+                    # Add sample ancestry for the currently-processed segment set.
+                    self.add_ancestry(input_id, left, right, ancestry_node)
+            prev_right = right
+
+        if is_sample and prev_right != self.sequence_length:
+            # If a trailing gap exists in the sample ancestry, fill it in.
+            self.add_ancestry(input_id, prev_right, self.sequence_length, output_id)
+        if output_id != -1:
+            num_edges = self.flush_edges()
+            if num_edges == 0 and not is_sample:
+                self.rewind_node(input_id, output_id)
+
+    def record_edge(self, left, right, parent, child):
+        """
+        Adds an edge to the output list.
+        """
+        if self.reduce_to_site_topology:
```

Comment:
I don't think this is relevant here, is it?

Reply:
Thanks, must have missed this - I think I've removed all other site/mutation related code.